### ISSUE TITLE
Uncaught TypeError: Cannot read property 'inline-0' of undefined

### ISSUE THREAD

**Issue by rkeeler** [keajs/kea#23]
I am running into this error on `v0.20.5` when trying to use the `connect` functionality, following the [example](https://kea.js.org/guide/connected) in the docs.

```js
Uncaught TypeError: Cannot read property 'inline-0' of undefined
```

Importing a logic from a separate file and applying it in its entirety to a component works fine. However, trying to connect specific actions or props to a component fails. I've tried connecting using a decorator, with the shorthand syntax (`connect(options)`), and with the long syntax (`kea({ connect: { ... } })`, but i get the same error in all cases.

I am able to reproduce this in a [simple Gist](https://gist.github.com/rkeeler/82b0670db66716336c3902ab6f6c2f3b). Is there a step that i'm missing to make `connect()` work or is this a bug in kea? I noticed that there has been some recent development activity, so maybe the docs are just outdated.

Please let me know if there is anything you need me to clarify about what is happening. 
Thanks!

Edit: Here's the full stack trace.
```
selectors.js:6 Uncaught TypeError: Cannot read property 'inline-0' of undefined
    at concat.reduce (selectors.js:6)
    at Array.reduce (<anonymous>)
    at pathSelector (selectors.js:6)
    at rootSelector (selectors.js:22)
    at index.js:86
    at Object.reducer1 (index.js:36)
    at Object.keys.forEach.propKey (kea.js:317)
    at Array.forEach (<anonymous>)
    at kea.js:316
    at Object.runComponentSelector [as run] (connectAdvanced.js:26)
    at Connect.initSelector (connectAdvanced.js:178)
    at new Connect (connectAdvanced.js:119)
    at ReactCompositeComponent.js:294
    at measureLifeCyclePerf (ReactCompositeComponent.js:75)
    at ReactCompositeComponentWrapper._constructComponentWithoutOwner (ReactCompositeComponent.js:293)
    at ReactCompositeComponentWrapper._constructComponent (ReactCompositeComponent.js:279)
    at ReactCompositeComponentWrapper.mountComponent (ReactCompositeComponent.js:187)
    at Object.mountComponent (ReactReconciler.js:45)
    at ReactCompositeComponentWrapper.performInitialMount (ReactCompositeComponent.js:370)
    at ReactCompositeComponentWrapper.mountComponent (ReactCompositeComponent.js:257)
    at Object.mountComponent (ReactReconciler.js:45)
    at ReactCompositeComponentWrapper.performInitialMount (ReactCompositeComponent.js:370)
    at ReactCompositeComponentWrapper.mountComponent (ReactCompositeComponent.js:257)
    at Object.mountComponent (ReactReconciler.js:45)
    at mountComponentIntoNode (ReactMount.js:104)
    at ReactReconcileTransaction.perform (Transaction.js:143)
    at batchedMountComponentIntoNode (ReactMount.js:126)
    at ReactDefaultBatchingStrategyTransaction.perform (Transaction.js:143)
    at Object.batchedUpdates (ReactDefaultBatchingStrategy.js:62)
    at Object.batchedUpdates (ReactUpdates.js:97)
    at Object._renderNewRootComponent (ReactMount.js:319)
    at Object._renderSubtreeIntoContainer (ReactMount.js:401)
    at Object.render (ReactMount.js:422)
    at Object../src/index.js (index.js:79)
    at __webpack_require__ (bootstrap 7abea5da1d167805669e:659)
    at fn (bootstrap 7abea5da1d167805669e:85)
    at Object.0 (index.js:80)
    at __webpack_require__ (bootstrap 7abea5da1d167805669e:659)
    at ./node_modules/ansi-regex/index.js.module.exports (bootstrap 7abea5da1d167805669e:708)
    at bundle.js:712
```

**Comment by mariusandra** [comment]
Hey, I believe the issue is that your code is running a call to `kea` before it has run `keaReducer`.
[This page](https://kea.js.org/api/reducer) in the docs should describe the issue.

Basically change your code to look like this:

```js
const scenesReducer = keaReducer('scenes');
const store = finalCreateStore(
  combineReducers({
    scenes: scenesReducer,
  })
);
```

That said, the error message should definitely be improved. So if this fixes your problem, please leave the issue open so I'll remember to improve the error message :).

**Comment by rkeeler** [comment]
Thanks for responding, but unfortunately the fix you suggested did not work. I posted an [updated Gist](https://gist.github.com/rkeeler/82b0670db66716336c3902ab6f6c2f3b) with the change you suggested. I'm calling `keaReducer('scenes')` on line 15 and my first call to `kea()` isn't until line 23. I'd also added line 82 to show that with the same setup that `logic1(logic2(App))` works but `connect({...})(App)` doesn't.

Do you have any other ideas?
Thanks!

**Comment by mariusandra** [comment]
Hey! Thank you for the gist, it was really helpful!

I'm investigating the issue and will try to come up with a proper solution. Until then I have a temporary fix that will solve the problem.

The fix is to add `store.dispatch({type: ''})` before the line that calls `ReactDOM.render()`.

I even had a [TODO](https://github.com/keajs/kea/blob/master/src/logic/kea.js#L312) in the code to investigate and fix this.

The issue is that whenever `kea` adds any logic to the redux store, it doesn't automatically regenerate the store itself. This is mainly for performance reasons, e.g. if you have hundreds of components that want to add themselves to the store... although this might not be as big of an issue as I'm making it out to be. I'll update this issue once a proper fix is in place!

**Comment by mariusandra** [comment]
Hey, can you please try if `v0.20.8` fixes the issue for you (without the temporary fix from my last comment)? It seems to have done the trick for me.

**Comment by rkeeler** [comment]
I updated to `v0.20.8` but that seems to have exposed a different issue.
```
store.js:26 Uncaught ReferenceError: regeneratorRuntime is not defined
    at Object../node_modules/kea/lib/scene/store.js (store.js:26)
    at __webpack_require__ (bootstrap 94e669b7222330b081ce:659)
    at fn (bootstrap 94e669b7222330b081ce:85)
    at Object../node_modules/kea/lib/logic/props.js (props.js:22)
    at __webpack_require__ (bootstrap 94e669b7222330b081ce:659)
    at fn (bootstrap 94e669b7222330b081ce:85)
    at Object../node_modules/kea/lib/logic/kea.js (kea.js:9)
    at __webpack_require__ (bootstrap 94e669b7222330b081ce:659)
    at fn (bootstrap 94e669b7222330b081ce:85)
    at Object../node_modules/kea/lib/index.js (index.js:8)
    at __webpack_require__ (bootstrap 94e669b7222330b081ce:659)
    at fn (bootstrap 94e669b7222330b081ce:85)
    at Object../src/index.js (fetch.js:461)
    at __webpack_require__ (bootstrap 94e669b7222330b081ce:659)
    at fn (bootstrap 94e669b7222330b081ce:85)
    at Object.0 (index.js:86)
    at __webpack_require__ (bootstrap 94e669b7222330b081ce:659)
    at ./node_modules/ansi-regex/index.js.module.exports (bootstrap 94e669b7222330b081ce:708)
    at bundle.js:712
./node_modules/kea/lib/scene/store.js @ store.js:26
__webpack_require__ @ bootstrap 94e669b7222330b081ce:659
fn @ bootstrap 94e669b7222330b081ce:85
./node_modules/kea/lib/logic/props.js @ props.js:22
__webpack_require__ @ bootstrap 94e669b7222330b081ce:659
fn @ bootstrap 94e669b7222330b081ce:85
./node_modules/kea/lib/logic/kea.js @ kea.js:9
__webpack_require__ @ bootstrap 94e669b7222330b081ce:659
fn @ bootstrap 94e669b7222330b081ce:85
./node_modules/kea/lib/index.js @ index.js:8
__webpack_require__ @ bootstrap 94e669b7222330b081ce:659
fn @ bootstrap 94e669b7222330b081ce:85
./src/index.js @ fetch.js:461
__webpack_require__ @ bootstrap 94e669b7222330b081ce:659
fn @ bootstrap 94e669b7222330b081ce:85
0 @ index.js:86
__webpack_require__ @ bootstrap 94e669b7222330b081ce:659
./node_modules/ansi-regex/index.js.module.exports @ bootstrap 94e669b7222330b081ce:708
(anonymous) @ bundle.js:712
```

I went back and tried some of the different versions with the same Gist as before. This new error appears to have been introduced in `v0.20.6`, so i'm not sure if it's actually related to the original error we have been discussing. I can open a new issue if you would like. However, i'm not able to verify the fix you added in `v0.20.8`

**Comment by mariusandra** [comment]
Hey, try installing and enabling `babel-polyfill`, either by adding `import
'babel-polyfill'` to the top of the file or adding it to your webpack conf.

**Comment by mariusandra** [comment]
Hey, try installing and enabling `babel-polyfill`, either by adding `import 'babel-polyfill'` to the top of the file or adding it to your webpack conf.

See here fot more info: http://babeljs.io/docs/usage/polyfill/

**Comment by rkeeler** [comment]
It's interesting that something changed between versions 0.20.5 and 0.20.6, but I added `babel-polyfill` to my webpack config and it fixed that issue. And i was able to test the original issue i reported with `connect()`. It works now!

Thanks for the quick turnaround in diagnosing the issue and releasing a fix!

**Comment by rkeeler** [comment]
As requested, i'm leaving the issue open so you remember to add a friendly error message instead of `Uncaught TypeError: Cannot read property 'inline-0' of undefined`. Feel free to close this issue when you see fit.

**Comment by mariusandra** [comment]
Hey, great to hear that it works! I wanted to write a longer reply yesterday, but couldn't due to time and device constraints.

There was something that changed between `0.20.5` and `0.20.6` which might feel like a breaking change and perhaps I didn't manage in the best way. What happened was that from `0.19.12` to `0.20.5`, I had added a `module` field in `package.json` pointing to the untranspiled ES6+ source of the library. I thought this was the right way to do it, but it resulted in various bugs with bundling your app. In short, this left kea completely untranspiled, and when you opened your app in a non-latest-chrome browser, everything broke, as almost nobody re-transpiles what comes to them through `node_modules`.

What other libraries, like `redux-saga` do, is to still transpile your app to ES5 for the paths that are referenced with `module`, but keep the `import` statements so that webpack 2+, rollup, etc can properly do tree shaking on them. I will add this change when I release `v0.21`.

That said, in reality the change between `0.20.5` and `0.20.6` means that you just got the bug sooner. You would probably have gotten the same thing when you would have written any code that uses sagas.

I'll keep the issue open until I have time to see what I can do with the error and update the documentation about `redux-saga` needing `babel-polyfill`.

Thanks for your patience! :)

**Comment by patrick-jones** [comment]
I believe you could use the `esnext: src/index.js` property in `package.json` to point at the untranspiled source. Here's an article that explains it: http://2ality.com/2017/06/pkg-esnext.html

The good thing about the approach used in the article (as I understand it) is that the library consumer has to opt in to using the untranspiled code. People who don't care will get the default, transpiled behavior.

Personally I would much prefer to transpile kea myself, as it is nearly impossible to debug kea-related errors if you have to deal with the regeneratorRuntime code. The only way I was able to debug #32 was to clone the repo, `yarn link` to it and change the `main` entry in `package.json` to point to `src/index.js`

**Comment by mariusandra** [comment]
Hi, `0.24.1` is out where:
- I now again have a version with `import`s instead of `require`s through the `"module"` field. This decreased my uncompressed bundle size by about 10KB. 
- I added the "esnext" field as suggested by @patrick-jones 

Closing this issue now. Feel free to reopen if there are any issues with the above

**Comment by k1sul1** [comment]
@mariusandra Ran into this early on trying to setup a new create-react-app project. I'm created a barebones example here: https://github.com/k1sul1/kea-bug-hunt

As you can see, store.dispatch({ type: '' }) didn't do much. 

![image](https://user-images.githubusercontent.com/2719615/42414436-26891b44-823e-11e8-9099-9ced1ff4e199.png)

**Comment by k1sul1** [comment]
Found it. Took some time off and did a docs compare. 

https://github.com/k1sul1/kea-bug-hunt/blob/master/src/index.js#L7
https://kea.js.org/guide/installation

It says to import `store` _before_ `App`. I guess it wasn't kidding.
